[Feature] Enable `kwargs` for single component models (via `extra`) (#1032)
mckinsey/vizro: vizro-core/docs/pages/explanation/schema.md

# Vizro schema and grammar of dashboards

> This post aims to explain the Vizro JSON schema, our attempt to create a grammar of dashboards and the role of Pydantic.

Reviewer: Aha, I spotted a section below. You can just link to that!

## What is a JSON schema?

There are many explanations out there explaining [what a JSON schema is](https://blog.postman.com/what-is-json-schema) - so here is only a very short intro.

> A JSON schema defines the rules, structure, and constraints that JSON data (a popular data format, used widely on the web) should follow, which leaves minimal room for assumptions and makes it a lot easier to predict the nature and behavior of JSON data.

Reviewer: ```suggestion
A JSON schema defines the rules, structure, and constraints that JSON data (a popular data format, used widely on the web) should follow. Use of a schema leaves minimal room for assumptions and makes it easier to predict the nature and behavior of JSON data.
```

> A JSON schema defines the rules, structure, and constraints that JSON data (a popular data format, used widely on the web) should follow. Use of a schema leaves minimal room for assumptions and makes it easier to predict the nature and behavior of JSON data.

Reviewer 2: ```suggestion
A [JSON schema](https://json-schema.org/) defines the rules, structure, and constraints that JSON data (a popular data format, used widely on the web) should follow. Use of a schema leaves minimal room for assumptions and makes it easier to predict the nature and behavior of JSON data.
```

An example of a JSON schema would be

```json
{
  "properties": {
    "A": {
      "title": "A",
      "type": "integer"
    },
    "B": {
      "items": {
        "type": "string"
      },
      "title": "B",
      "type": "array"
    }
  },
  "required": [
    "A",
    "B"
  ],
  "title": "Example",
  "type": "object"
}
```

> This schema tells us that whenever we receive JSON data, and want to check whether it is valid, the data needs to contain two fields - a field `A`, which must be a single integer, and a field `B`, which must be an array of strings. Both fields need to be provided - they are required.

Reviewer: Maybe use "Both fields are mandatory -- they are required" ??

> In practice this means that the following data would be valid or invalid.

Author: I think i liked it better before because it took less space and flicking between valid and invalid made it easier to spot the difference. Wdyt?

=== "valid"
    ```json
    {
      "A": 1,
      "B": [
        "a",
        "b"
      ]
    }
    ```

> === "invalid"

Reviewer 2: ```suggestion
=== "invalid (`B` is not an array of strings)"
```

    ```json
    {
      "A": 1,
      "B": [
        "a",
        3
      ]
    }
    ```

> === "also invalid"

Reviewer 2: ```suggestion
=== "invalid (required field B is missing)"
```

    ```json
    {

>     "A": 1,

Reviewer 2: ```suggestion
      "A": 1,
```

    }
    ```

## What is the Vizro JSON schema?

Similar to the above example, the Vizro framework also has a JSON schema. It can be [found in our GitHub repository](https://github.com/mckinsey/vizro/tree/main/vizro-core/schemas). It is more complicated than the simple schema above, but it generally follows the same principle.

> 

Reviewer 2: Is it possible here to add another simple example that more closely matches the above toy example?

i.e. can you extract a very small piece of the vizro schema, write it out explicitly here (even if it means doing "..." or removing some extra bits like `description`), and then provide valid and invalid configurations for that schema?

I think this would make it much clearer how the Vizro schema is analogous to the above example.

If that's all too much work then maybe provide an example of an invalid configuration for the below example, e.g. `title` is not defined or something is the wrong type.

Author: I made a fake simplifed schema, anything beyond that I think is too much!

This means you can configure a Vizro dashboard according to a set of constraints that are defined in the schema. The configuration language that you choose is secondary - it can be via Python, but also via JSON or YAML. Exactly this is shown in [our showcase of configuration options](../user-guides/dashboard.md#use-dashboard-configuration-options).

> === "This json..."

Reviewer: ```suggestion
=== "This JSON..."
```

    ```json
    {
      "pages": [
        {
          "components": [
            {
              "figure": {
                "_target_": "scatter",
                "color": "species",
                "data_frame": "iris",
                "x": "sepal_length",
                "y": "petal_width"
              },
              "type": "graph"
            },
            {
              "figure": {
                "_target_": "histogram",
                "color": "species",
                "data_frame": "iris",
                "x": "sepal_width"
              },
              "type": "graph"
            }
          ],
          "controls": [
            {
              "column": "species",
              "type": "filter"
            }
          ],
          "title": "My first dashboard"
        }
      ]
    }
    ```

=== "maps to this Dashboard"
    [![Dashboard]][dashboard]

!!! note

>     The Vizro schema is not yet complete. This means that it does not yet define everything that we consider a core and supported functionality of Vizro. The most prominent example of this is the [`CapturedCallable`][vizro.models.types.CapturedCallable] - the objects you insert into many models such as `vm.Graph`, `vm.Table` or `vm.Action`, often into the `figure` argument.

Reviewer: ```suggestion
    The Vizro schema is still incomplete. This means that it does not yet define everything that we consider to be core and supported functionality of Vizro. The most prominent example of an omission is the [`CapturedCallable`][vizro.models.types.CapturedCallable], which are the objects you insert into many models such as `vm.Graph`, `vm.Table` or `vm.Action`, often into the `figure` argument.
```

>     This does not mean that it cannot be configured in JSON - in fact, in the config above we have used the following JSON:

Reviewer: ```suggestion
    This does not mean that it cannot be configured in JSON. In fact, in the configuration above, we have used the following JSON:
```

    ```json
    {
      "figure": {
        "_target_": "histogram",
        "color": "species",
        "data_frame": "iris",
        "x": "sepal_width"
      },
      "type": "graph"
    }
    ```

>     It does mean however that e.g. `figure` does not appear in the schema - and as such is not clearly defined as part of the Vizro framework, although it should. We are working on completing the Vizro schema in the future.

Reviewer: ```suggestion
    It means that `figure` does not appear in the schema so, as such, it is not clearly defined as part of the Vizro framework, although it should be considered as such. We are working to complete the Vizro schema in the future.
```

Reviewer: Is this kind of task (completing the schema) something we could offer as a "help wanted" ticket on the public repo for a contribution? If so, do you want to make an issue and link it in here?

Author: It would be great if we received help on that, but I think it is a little too complicated to be labelled a "help wanted" ticket, or at least until we have defined a little more the directionality

## The role of Pydantic

> The Vizro framework is powered by [Pydantic](https://docs.pydantic.dev/latest/) - the most widely used data validation library for Python. Pydantic is acts as the glue that connects Vizro's JSON schema to it's emergent manifestation - a React front end served by a Flask backend, facilitated by Dash.

Reviewer: ```suggestion
The Vizro framework is powered by [Pydantic](https://docs.pydantic.dev/latest/), which is the most widely used data validation library for Python. Pydantic acts as the glue that connects Vizro's JSON schema to its emergent manifestation - a React front end served by a Flask backend, facilitated by Dash.
```

Reviewer: "emergent manifestation" sounds a bit scary to me 👻 and to be honest, I don't know what it means in this context. Is there a way to say this more simply?

Author: Haha yes, good point!

> The Vizro framework is powered by [Pydantic](https://docs.pydantic.dev/latest/), which is the most widely used data validation library for Python. Pydantic acts as the glue that connects Vizro's JSON schema to its actual implementation - a React front end served by a Flask backend, facilitated by Dash.

Reviewer 3: ```suggestion
The Vizro framework is powered by [Pydantic](https://docs.pydantic.dev/latest/), which is the most widely used data validation library for Python. Pydantic acts as the glue that connects Vizro's JSON schema to its actual implementation - a React frontend served by a Flask backend, facilitated by Dash.
```

> One core advantage of Pydantic is that it emits a [JSON schema](https://blog.postman.com/what-is-json-schema) - so that our models such as `vm.Page` or `vm.Filter` can easily be translated into a well-defined JSON schema. Beyond this capability, it also has support for custom validation beyond the functionality of the JSON schema.

Reviewer: ```suggestion
One core advantage of Pydantic is that it emits a [JSON schema](https://blog.postman.com/what-is-json-schema) so that our models, such as `vm.Page` or `vm.Filter`, can easily be translated into a well-defined JSON schema. Beyond this capability, it also has support for custom validation beyond the functionality of the JSON schema.
```

> One core advantage of Pydantic is that it emits a [JSON schema](https://blog.postman.com/what-is-json-schema) so that our models, such as `vm.Page` or `vm.Filter`, can easily be translated into a well-defined JSON schema. Beyond this capability, it also has support for custom validation beyond the functionality of the JSON schema.

Reviewer 2: ```suggestion
One core advantage of Pydantic is that it [can automatically produce a JSON schema](https://docs.pydantic.dev/latest/concepts/json_schema/) from our models (such as `vm.Page` or `vm.Filter`). Pydantic also supports custom validation beyond the functionality of JSON schema.
```

> Sticking with our toy example from above - we could for example define that in the array of strings, consecutive elements have to have more letters than their predecessor. This is of course fairly random, but it shows the power of custom validation for anyone using the Vizro framework. Most of the time when the user "misconfigures" - the error message is clear and concise.

Reviewer: ```suggestion
In our toy example from above we could define that, in the array of strings, consecutive elements have to have more letters than their predecessor. This is a fairly random example, but it illustrates the power of custom validation for anyone using the Vizro framework. Most of the time when the user "misconfigures", the error message is clear and concise.
```

> In our toy example from above we could define that, in the array of strings, consecutive elements have to have more letters than their predecessor. This is a fairly random example, but it illustrates the power of custom validation for anyone using the Vizro framework. Most of the time when the user "misconfigures", the error message is clear and concise.

Reviewer 2: ```suggestion
In our [toy example above](#what-is-a-json-schema) we could define that, in the array of strings, consecutive elements have to have more letters than their predecessor. This is a fairly random example, but it illustrates the power of custom validation for anyone using the Vizro framework. Most of the time when the user "misconfigures", the error message is clear and concise.
```

Reviewer 2: This doesn't quite make sense to me. "we could define that ..." - is that something you can do in pure JSON schema or pydantic itself enables?

Can you create the pydantic model that would output the toy example JSON?

I feel this would be a bit more compelling if the narrative went like:
* pydantic model to generate toy example JSON + valid/invalid examples + ValidationError raised
* pydantic model that extends that and has a constraint that can be encoded in JSON (regex or something) + valid/invalid examples + ValidationError raised
* pydantic model that extends that and has a constraint that can't be encoded in JSON + valid/invalid examples + ValidationError raised

Author: I did it similar, but a bit shorter. Let me know what you think

## The grammar of dashboards

> How does this all come together? One of the long-term goals of Vizro is to define a so-called **grammar of dashboards**. This means that we complete the Vizro schema and define a unified, implementation independent language to configure dashboards.

Reviewer 2: This section should definitely link to other grammars:
* grammar of graphics
* grammar of interactive graphics
* plotly JSON schema

These are ultimately what inspires the term "grammar of dashboards" and roughly what we seek to emulate, even if we can never fully realise it to the same extent as them.

> At the moment, the Vizro framework serves [Dash](https://github.com/plotly/dash) apps - but this is not a necessary condition. In principle, the mapping of JSON to application/dashboard could be realized in any technology.

Reviewer 2: ```suggestion
At the moment, the Vizro framework serves [Dash](https://github.com/plotly/dash) apps — but this is not a necessary condition. In principle, the mapping of JSON to app could be realized using other technologies.
```

(Yes, I am nitpicky enough to lengthen your hyphen into a dash.)

## The role of `extra`

> Some of our models, e.g. the [`Container`][vizro.models.Container], have an argument called `extra`. This argument allows the user to pass extra arguments directly to the underlying component of the model. In the case of the `Container` this would be the [`dbc.Container`](https://dash-bootstrap-components.opensource.faculty.ai/docs/components/layout/).

Reviewer: ```suggestion
Some of our models, e.g. the [`Container`][vizro.models.Container], have an argument called `extra`. This argument enables the user to pass extra arguments directly to the underlying component of the model. In the case of the `Container`, this would be the [`dbc.Container`](https://dash-bootstrap-components.opensource.faculty.ai/docs/components/layout/).
```

> This is clearly implementation specific, and as such, the `extra` argument of on purpose excluded from the Vizro schema. The argument helps users to get the maximum flexibility quickly without having to go beyond the Vizro framework, but it should not be assumed to be a core part of Vizro. As such, this argument may break in future releases of Vizro.

Reviewer: ```suggestion
This is clearly implementation specific and, as such, the `extra` argument is (on purpose) excluded from the Vizro schema. The argument helps users to get the maximum flexibility quickly without having to go beyond the Vizro framework, but it should not be assumed to be a core part of Vizro. This argument may break in future releases of Vizro.
```

Reviewer: Instead of "break" do you want to say "be removed"? "Deprecated" ?

Author: Hm no, I guess it's more like: Using this argument may break your code in future releases of Vizro
